Modules/ManualAssessment/classes/Members/class.ilManualAssessmentMembers.php

```diff
@@ -0,0 +1,226 @@
+
+class ilManualAssessmentMembers implements Iterator, Countable {
+	protected $member_records = array();
+	protected $position = 0;
+	protected $mass;
+
+	const FIELD_FIRSTNAME = 'firstname';
+	const FIELD_LASTNAME = 'lastname';
+	const FIELD_LOGIN = 'login';
+	const FIELD_USR_ID = 'usr_id';
+	const FIELD_LEARNING_PROGRESS = 'learning_progress';
+	const FIELD_EXAMINER_ID = 'examiner_id';
+	const FIELD_EXAMINER_FIRSTNAME = 'examiner_firstname';
+	const FIELD_EXAMINER_LASTNAME = 'examiner_lastname';
+	const FIELD_RECORD = 'record';
+	const FIELD_INTERNAL_NOTE = 'internal_note';
+	const FIELD_NOTIFY = 'notify';
+	const FIELD_FINALIZED = 'finalized';
+	const FIELD_NOTIFICATION_TS = 'notification_ts';
+
+	const LP_IN_PROGRESS = ilLPStatus::LP_STATUS_IN_PROGRESS_NUM;
+	const LP_COMPLETED = ilLPStatus::LP_STATUS_COMPLETED_NUM;
+	const LP_FAILED = ilLPStatus::LP_STATUS_FAILED_NUM;
+
+	public function __construct(ilObjManualAssessment $mass) {
+		$this->mass = $mass;
+	}
+
+	/**
+	 * Countable Methods 
+	 */
+	public function count() {
+		return count($this->member_records);
+	}
+
+	/**
+	 * Iterator Methods
+	 */
+	public function current() {
+		return current($this->member_records);
+	}
+
+	public function key() {
+		return key($this->member_records);
+	}
+
+	public function next() {
+		$this->position++;
+		next($this->member_records);
+	}
+
+	public function rewind() {
+		$this->position = 0;
+		reset($this->member_records);
+	}
+
+	public function valid() {
+		return $this->position < count($this->member_records);
+	}
+
+	/**
+	 * Get the manual assessment object that is corresponding to this
+	 *
+	 * @return	ilObjManualAssessment
+	 */
+	public function referencedObject() {
+		return $this->mass;
+	}
+
+	/**
+	 * Check the validity of a record before adding it to this
+	 *
+	 * @param int|string|null[]	$record
+	 * @return	bool
+	 */
+	public function recordOK(array $record) {
+		if(isset($record[self::FIELD_USR_ID])) {
+			if(!$this->userExists($record[self::FIELD_USR_ID])
+				|| $this->userAllreadyMemberByUsrId($record[self::FIELD_USR_ID])) {
+				var_dump($record);
+				die();
+				return fasle;
+			}
+		}
+		if(!in_array($record[self::FIELD_LEARNING_PROGRESS],
+			array(self::LP_FAILED, self::LP_COMPLETED, self::LP_IN_PROGRESS))) {
+			var_dump($record);
+			die();
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Check if a user with user_id is member of this
+	 *
+	 * @param int|string	$usr_id
+	 * @return	bool
+	 */
+	public function userAllreadyMemberByUsrId($usr_id) {
+		return isset($this->member_records[$usr_id]);
+	}
+
+	/**
+	 * Check if a user is member of this
+	 *
+	 * @param ilObjUser	$usr
+	 * @return	bool
+	 */
+	public function userAllreadyMember(ilObjUser $usr) {
+		return $this->userAllreadyMemberByUsrId($usr->getId());
+	}
+
+	protected function userExists($usr_id) {
+		return ilObjUser::_exists($usr_id,false,'usr');
+	}
+
+	/**
+	 * Clone this and add an additional record
+	 *
+	 * @param	int|string|null[]	$record
```

Comment:
Is this a list where each field is either a int, string, or null? This, to me, would be (int|string|null)[]. The given form is ambiguous.